src/main/kotlin/com/atlassian/performance/tools/jiraactions/api/memories/adaptive/AdaptiveJqlMemory2.kt

```diff
@@ -0,0 +1,68 @@
+package com.atlassian.performance.tools.jiraactions.api.memories.adaptive
+
+import com.atlassian.performance.tools.jiraactions.api.SeededRandom
+import com.atlassian.performance.tools.jiraactions.api.jql.Jql
+import com.atlassian.performance.tools.jiraactions.api.jql.JqlSupplier
+import com.atlassian.performance.tools.jiraactions.api.memories.JqlMemory
+import com.atlassian.performance.tools.jiraactions.api.memories.JqlMemory2
+import com.atlassian.performance.tools.jiraactions.api.memories.adaptive.jql.JqlPrescription
+import com.atlassian.performance.tools.jiraactions.api.memories.adaptive.jql.JqlPrescriptions
+import com.atlassian.performance.tools.jiraactions.api.page.IssuePage
+import com.atlassian.performance.tools.jiraactions.jql.BuiltInJQL
+import com.atlassian.performance.tools.jiraactions.jql.JqlContextImpl
+import org.apache.logging.log4j.LogManager
+import org.apache.logging.log4j.Logger
+import java.util.function.Predicate
+
+class AdaptiveJqlMemory2(
```

Comment:
Why this class is public? We have already public interface JqlMemory2.
Should we create some sort of factory that returns the instance of JqlMemory2 in this case AdaptiveJqlMemory2?
What should be a general idea here?

Reply:
in general classes with "2" suffix are replacements, for existing which will be deprecated and removed with major change, so in some time number of interfaces will be the same as before refactoring

Reply:
@lipinskipawel for more invasive refactoring You can look here
https://github.com/drspartez/jira-actions/tree/issue/memories-refactoring

Reply:
My question was more about the Factory class, I do understand the "2" suffix.
Anyway, good to see Factory in the commit you sent.

Reply:
OK. So concern of this change is to replace "String" jql representation with something else which is more relevant (so all JQL interfaces ore for this reason). Other stuff remains almost untouched (so we still offer Adaptive memories implementations as usual). 
In second approach (https://github.com/drspartez/jira-actions/tree/issue/memories-refactoring) i changed the fashion we provide adaptive memories (hidden implementation and factories), also reduced number of actual memory interfaces to 2.

Reply:
> Why this class is public?

The old `AdaptiveJqlMemory` was also API, so it's perfectly fine for the replacement to be public as well.
It was published to make custom scenarios easier to write.

Reply:
> Should we create some sort of factory that returns the instance of JqlMemory2 in this case AdaptiveJqlMemory2?

Factories cost complexity and must have additional value to be justified. AFAIK across all JPT libs we expose only one factory: `com.atlassian.performance.tools.report.api.result.RawCohortResult.Factory`. We did that to offer different kinds of `RawCohortResult` without allowing custom subtypes.

Why did we forbid custom subtypes? Because we needed to add methods to the (old) `CohortResult`, but we didn't have a pattern for that. So we created a replacement (just like `JqlMemory2`) and we wanted to protect ourselves from this problem in the future. We decided nobody should need to implement this interface. In short, we wanted `RawCohortResult` to be API, but not SPI. 

The question is - do we want to allow custom implementations of `JqlMemory2`? If so, let's go for a factory, else public constructors/builders.